### PR TITLE
Restructure migration code to have less UX integration

### PR DESCRIPTION
**Pull request by robbrockbank**
## Description

Remove as many traces of UX from the migration code (before we move it to libcalico-go).  In particular:
-  No mention of UX specific commands should be in the migration code.  Text should be generic.  Especially no switches etc.
-  No user interaction in the UX
-  No printing to stdout in the migration code (now handled by an optional status interface)

Also improves the error handling by tracing the last error in the final report message.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

File: calico_upgrade/pkg/migrate/migrate.go [hunk 1]
```diff
@@ -45,33 +46,54 @@ const (
...
+	ValidateConversion() (*MigrationData, error)
+	IsDestinationEmpty() (bool, error)
+	Migrate() (*MigrationData, error)
+	Abort() error
+	Complete() error
+}
 
-func DisplayStatusMessages(d bool) {
-	displayStatus = d
+type StatusWriterInterface interface {
```
Comment (bcreane) [inline calico_upgrade/pkg/migrate/migrate.go]: `MigrationStatusWriterInterface`

File: calico_upgrade/pkg/commands/clihelper.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+import (
+	"fmt"
+	"os"
+	"strings"
+)
+
+const (
+	lineLength = 79
+	sepChar    = '*'
```
[the remark this answers is not in the document]
Comment (robbrockbank) [reply calico_upgrade/pkg/commands/clihelper.go]: kk - i was on the fence!

File: calico_upgrade/pkg/commands/clihelper.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+		sep[i] = sepChar
+	}
+	m.NewLine()
+	fmt.Println(string(sep))
+	m.NewLine()
+}
+
+// Display a generic proceed? message and wait for a "yes" response.
+func (m *cliHelper) ConfirmProceed() {
+	fmt.Print("Type yes to proceed (any other input cancels): ")
```
[the remark this answers is not in the document]
Comment (robbrockbank) [reply calico_upgrade/pkg/commands/clihelper.go]: ah yes - my bad i meant to change that

File: calico_upgrade/pkg/commands/dryrun.go [hunk 4]
```diff
@@ -78,43 +78,74 @@ Description:
...
-	if res == migrate.ResultOK {
-		// We validated the data successfully. Include a report.
-		printFinalMessage("Successfully validated v1 to v3 conversion.\n" +
-			"See reports below for details of the conversion.")
-		printAndOutputReport(output, data)
+	// Perform the validation.
+	data := validate(m, ch, output, ignoreV3Data)
+	if len(data.Resources) == 0 {
+		// For non-KDD this will be an error case and so we won't hit this.
+		ch.Msg("There is no data requiring conversion.  You may proceed with the upgrade without " +
```
[the remark this answers is not in the document]
Comment (robbrockbank) [reply calico_upgrade/pkg/commands/dryrun.go]: lalalala.  It's such a baked in habit now

**Pull request merged**